force-app/main/default/classes/Rollups.cls

```diff
@@ -0,0 +1,187 @@
+        Contact.SObjectType => ServiceDelivery__c.Contact__c,
+        ProgramEngagement__c.SObjectType => ServiceDelivery__c.ProgramEngagement__c,
+        ServiceSession__c.SObjectType => ServiceDelivery__c.ServiceSession__c,
+        Service__c.SObjectType => ServiceDelivery__c.Service__c
+    };
+
+    private ServiceDeliveryService deliveryService = new ServiceDeliveryService();
+    private SystemAccess systemAccess = new SystemAccess();
+    private Map<Id, RollupRecord> rollupRecordById = new Map<Id, RollupRecord>();
+    private Map<Id, RollupRecord> processedRecordsById = new Map<Id, RollupRecord>();
+    private SObjectType sObjectType;
+    private SObjectField lookupField;
+
+    public Rollups(SObjectType sObjectType, SObjectField lookupField) {
+        this.sObjectType = sObjectType;
+        this.lookupField = lookupField;
+    }
+
+    public Set<Id> getProcessedRecordIds() {
+        return processedRecordsById.keySet();
+    }
+
+    public void rollupDeliveryRecords(List<ServiceDelivery__c> deliveries) {
+        if (deliveries.isEmpty()) {
+            return;
+        }
+
+        if (System.isBatch()) {
+            setTotalRecordCount(deliveries[0]);
+            setTotalRecordCount(deliveries[deliveries.size() - 1]);
+        }
+
+        for (ServiceDelivery__c delivery : deliveries) {
+            Id parentLookupId = (Id) delivery.get(lookupField);
+
+            if (!rollupRecordById.containsKey(parentLookupId)) {
+                RollupRecord record = createRollupRecord(parentLookupId);
+                rollupRecordById.put(parentLookupId, record);
+            }
+
+            RollupRecord parentRecord = rollupRecordById.get(parentLookupId);
+            incrementRollup(delivery, parentRecord);
+        }
+
+        for (Id rollupRecordId : rollupRecordById.keySet()) {
+            if (rollupRecordById.get(rollupRecordId).getIsProcessed()) {
+                processedRecordsById.put(
+                    rollupRecordId,
+                    rollupRecordById.remove(rollupRecordId)
+                );
+            }
+        }
+    }
+
+    private void setTotalRecordCount(ServiceDelivery__c delivery) {
+        Id recordId = (Id) delivery.get(lookupField);
+        if (rollupRecordById.containsKey(recordId)) {
+            return;
+        }
+
+        RollupRecord record = createRollupRecord(recordId);
+        Integer totalRecords = Database.countQuery(
+            deliveryService.getDeliveryRollupsCountQuery(
+                String.valueOf(lookupField),
+                recordId
+            )
+        );
+        record.setTotalRecords(totalRecords);
+        rollupRecordById.put(recordId, record);
+    }
+
+    public void commitRollups() {
+        List<SObject> recordsToUpdate = new List<SObject>();
+        Set<Id> recordIds = processedRecordsById.keySet();
+        String query = buildOriginalRecordsQuery('Id IN: recordIds');
+
+        List<SObject> originalRecords = Database.query(query);
```

Comment:
Let's move the orignalRecordsQueryBuilder and query over to attendance selector and ask servicedeliveryservice for the records.